cas/ctg_store.c

```diff
@@ -1883,6 +2080,289 @@ static const struct m0_be_btree_kv_ops cas_btree_ops = {
+	struct m0_be_tx           *tx     = &ctg_op->co_fom->fo_tx.tx_betx;
+	struct m0_crv              new_version = M0_CRV_INIT_NONE;
+	struct m0_crv              old_version = M0_CRV_INIT_NONE;
+	int                        rc;
+
+	M0_PRE(ctg_op->co_is_versioned);
+	M0_ENTRY();
+
+	ctg_op_version_get(ctg_op, &new_version);
+	M0_ASSERT_INFO(!m0_crv_is_none(&new_version),
+		       "Versioned PUT or DEL without a valid version?");
+
+	/*
+	 * TODO: Performance.
+	 * This lookup call can be avoided if the version comparison
+	 * was shifted into btree. However, at this moment the cost of
+	 * such a lookup is much lower than the cost of re-working btree API.
+	 * See the be_btree_search() call inside ::btree_save. It is the
+	 * point where btree_save does its own "lookup". Versions could
+	 * be compared right in there. Alternatively, btree_save could
+	 * use btree_cursor as a "hint".
+	 */
+	rc = M0_BE_OP_SYNC_RET(op,
+			       m0_be_btree_lookup_inplace(btree,
+							  &op,
+							  key,
+							  anchor),
+			       bo_u.u_btree.t_rc) ?:
+		ctg_vbuf_unpack(&anchor->ba_value, &old_version);
+
+	/* The tree is long-locked anyway. */
+	m0_be_btree_release(NULL, anchor);
+
+	if (!M0_IN(rc, (0, -ENOENT)))
+		return M0_ERR(rc);
+
+	/*
+	 * Skip overwrite op if the existing record is "newer" (version-wise)
+	 * than the record to be inserted. Note, <= 0 means that we filter out
+	 * the operations with the exact same version and tombstone.
+	 */
+	if (!m0_crv_is_none(&old_version) &&
+	    m0_crv_cmp(&new_version, &old_version) <= 0)
+		return M0_RC(0);
+
+	M0_LOG(M0_DEBUG, "Overwriting " CRV_F " with " CRV_F ".",
+	       CRV_P(&old_version), CRV_P(&new_version));
+
+	anchor->ba_value.b_nob = ctg_vbuf_packed_size(&ctg_op->co_val);
+	rc = M0_BE_OP_SYNC_RET(op,
+			       m0_be_btree_save_inplace(btree, tx, &op,
+							key, anchor, true,
+							ctg_op_zones(ctg_op)),
+			       bo_u.u_btree.t_rc);
+
+	if (rc == 0) {
+		ctg_vbuf_pack(&ctg_op->co_anchor.ba_value,
+			      &ctg_op->co_val,
+			      &new_version);
+
+		m0_ctg_state_inc_update(tx,
+					key->b_nob -
+					sizeof(struct generic_key) +
+					ctg_op->co_val.b_nob);
+	}
+
+	return M0_RC(rc);
+}
+
+/*
+ * Gets an alive record (without tombstone set) in btree.
+ * Returns -ENOENT if tombstone is set.
+ * Always sets co_out_ver if the record physically exists in the tree.
+ */
+static int versioned_get_sync(struct m0_ctg_op *ctg_op)
+{
+	struct m0_be_btree        *btree  = &ctg_op->co_ctg->cc_tree;
+	struct m0_be_btree_anchor *anchor = &ctg_op->co_anchor;
+	int                        rc;
+
+	M0_PRE(ctg_op->co_is_versioned);
+
+	rc = M0_BE_OP_SYNC_RET(op,
+			       m0_be_btree_lookup_inplace(btree, &op,
+							  &ctg_op->co_key,
+							  anchor),
+			       bo_u.u_btree.t_rc) ?:
+		ctg_vbuf_unpack(&anchor->ba_value, &ctg_op->co_out_ver);
+
+	if (rc == 0) {
+		if (m0_crv_tbs(&ctg_op->co_out_ver))
+			rc = -ENOENT;
+		else
+			ctg_op->co_out_val = anchor->ba_value;
+	}
+
```

Comment:
Do we need to call  m0_be_btree_release(NULL, anchor); here?

In versioned_put_sync(), the same operation is called.
Why is it not called here in this function?

Reply:
versioned_put_sync reuses the same anchor:
```
// PUT:
anchor = ctg_op->op_anchor;
lookup_inplace(..., anchor, ...);
release(anchor);
save_inplace(..., anchor, ...);
// release(anchor); will be called by m0_ctg_op_fini
```
Here we do not reuse it:
```
// GET:
save_inplace(..., anchor, ...);
// release(anchor); will be called by m0_ctg_op_fini
```

Reply:
OK. I see.